### PR TITLE
Adjust num_embed to the Transformer model size.

### PR DESCRIPTION
**Pull request by tdomhan**
With this change the embedding size will be automatically adjusted to the Transformer model size in case they do not match instead of failing to train a model.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: sockeye/train.py [hunk 1]
```diff
@@ -368,7 +350,8 @@ def create_data_iters_and_vocabs(args: argparse.Namespace,
 def create_encoder_config(args: argparse.Namespace,
                           max_seq_len_source: int,
                           max_seq_len_target: int,
-                          config_conv: Optional[encoder.ConvolutionalEmbeddingConfig]) -> Tuple[encoder.EncoderConfig,
+                          config_conv: Optional[encoder.ConvolutionalEmbeddingConfig],
+                          num_embed_source: int) -> Tuple[encoder.EncoderConfig,
                                                                                                 int]:
```
Comment (artemsok) [inline sockeye/train.py]: indent?

Comment (tdomhan) [reply sockeye/train.py]: yeah the indentation was kind of weird. fixed.

**Comment by tdomhan** [comment]
Addressed offline comments by @davvil to only set num_embed to the Transformer model size if it is not specified by the user.

File: CHANGELOG.md [hunk 2]
```diff
@@ -10,6 +10,10 @@ Note that Sockeye has checks in place to not translate with an old model that wa
 
 Each version section may have have subsections for: _Added_, _Changed_, _Removed_, _Deprecated_, and _Fixed_.
 
+## [1.18.79]
+### Changed
+ - The embedding size is automatically adjusted to the Transformer model size in case they are not equal.
```
Comment (artemsok) [inline CHANGELOG.md]: add ""..if not specified explicitly"?

File: sockeye/train.py [hunk 3]
```diff
@@ -458,18 +440,19 @@ def create_encoder_config(args: argparse.Namespace,
...
+                          max_seq_len_source: int, max_seq_len_target: int,
+                          num_embed_target: int) -> decoder.DecoderConfig:
     """
     Create the config for the decoder.
 
     :param args: Arguments as returned by argparse.
     :param encoder_num_hidden: Number of hidden units of the Encoder.
     :param max_seq_len_source: Maximum source sequence length.
     :param max_seq_len_target: Maximum target sequence length.
+    :param num_embed_target: The size of the source embedding.
```
Comment (artemsok) [inline sockeye/train.py]: typo: target embedding

File: sockeye/train.py [hunk 4]
```diff
@@ -587,6 +570,48 @@ def check_encoder_decoder_args(args) -> None:
                         "Recurrent dropout without memory loss only supported for LSTMs right now.")
 
 
+def get_num_embed(args: argparse.Namespace) -> Tuple[int, int]:
+    num_embed_source, num_embed_target = args.num_embed
+    if args.encoder == C.TRANSFORMER_TYPE:
+        transformer_model_size_source = args.transformer_model_size[0]
+        if not num_embed_source:
+            logger.info("Source embedding size was not set t will automatically be adjusted to match the "
```
Comment (artemsok) [inline sockeye/train.py]: typo: t -> , it

File: sockeye/train.py [hunk 5]
```diff
@@ -587,6 +570,48 @@ def check_encoder_decoder_args(args) -> None:
...
+            check_condition(adjusted_transformer_encoder_model_size % 2 == 0 and
+                            adjusted_transformer_encoder_model_size % args.transformer_attention_heads[0] == 0,
+                            "Sum of source factor sizes, i.e. num-embed plus source-factors-num-embed, (%d) "
+                            "has to be even and a multiple of encoder attention heads (%d)" % (
+                                adjusted_transformer_encoder_model_size, args.transformer_attention_heads[0]))
+
+    if args.decoder == C.TRANSFORMER_TYPE:
+        transformer_model_size_target = args.transformer_model_size[1]
+        if not num_embed_target:
+            logger.info("Target embedding size was not set t will automatically be adjusted to match the "
```
Comment (artemsok) [inline sockeye/train.py]: same here

File: sockeye/train.py [hunk 6]
```diff
@@ -587,6 +570,48 @@ def check_encoder_decoder_args(args) -> None:
                         "Recurrent dropout without memory loss only supported for LSTMs right now.")
 
 
+def get_num_embed(args: argparse.Namespace) -> Tuple[int, int]:
+    num_embed_source, num_embed_target = args.num_embed
+    if args.encoder == C.TRANSFORMER_TYPE:
+        transformer_model_size_source = args.transformer_model_size[0]
+        if not num_embed_source:
+            logger.info("Source embedding size was not set t will automatically be adjusted to match the "
+                        "Transformer model size (%d).", transformer_model_size_source)
```
Comment (artemsok) [inline sockeye/train.py]: "Transformer _source_ model size"?

File: sockeye/train.py [hunk 7]
```diff
@@ -587,6 +570,48 @@ def check_encoder_decoder_args(args) -> None:
...
+                            adjusted_transformer_encoder_model_size % args.transformer_attention_heads[0] == 0,
+                            "Sum of source factor sizes, i.e. num-embed plus source-factors-num-embed, (%d) "
+                            "has to be even and a multiple of encoder attention heads (%d)" % (
+                                adjusted_transformer_encoder_model_size, args.transformer_attention_heads[0]))
+
+    if args.decoder == C.TRANSFORMER_TYPE:
+        transformer_model_size_target = args.transformer_model_size[1]
+        if not num_embed_target:
+            logger.info("Target embedding size was not set t will automatically be adjusted to match the "
+                        "Transformer model size (%d).", transformer_model_size_target)
```
Comment (artemsok) [inline sockeye/train.py]: "Transformer _target_ model size"?

**Comment by tdomhan** [comment]
rebased

**Pull request merged**